File: WinForms/CS/GridMainDemo/Modules/Custom Functions.cs
namespace DevExpress.XtraGrid.Demos {
    using System;
    using DevExpress.Data.Filtering;
    using DevExpress.Data.Summary;

    // Integer
    public class IsBlackFridayDiscountFunction : ICustomFunctionDisplayAttributes {
        public const string FunctionName = "IsBlackFridayDiscount";
        static readonly IsBlackFridayDiscountFunction Instance = new IsBlackFridayDiscountFunction();
        IsBlackFridayDiscountFunction() { }
        //
        public static void Register() {
            CriteriaOperator.RegisterCustomFunction(Instance);
        }
        public static bool Unregister() {
            return CriteriaOperator.UnregisterCustomFunction(Instance);
        }
        #region ICustomFunctionOperatorBrowsable Members
        public FunctionCategory Category {
            get { return FunctionCategory.Math; }
        }
        public string Description {
            get { return "The discount amount is 15% or more."; }
        }
        public bool IsValidOperandCount(int count) {
            return count == 1;
        }
        public bool IsValidOperandType(int operandIndex, int operandCount, Type type) {
            return SummaryItemTypeHelper.IsNumericalType(type);
        }
        public int MaxOperandCount {
            get { return 1; }
        }
        public int MinOperandCount {
            get { return 1; }
        }
        #endregion
        #region ICustomFunctionDisplayAttributes
        public string DisplayName {
            get { return "Is Black Friday Discount"; }
        }
        public object Image {
            get { return Properties.Resources.isBlackFridayDiscount; }
        }
        #endregion
        #region ICustomFunctionOperator Members
        public object Evaluate(params object[] operands) {
            double discount = Convert.ToDouble(operands[0]);
            return discount >= 0.15;
        }
        public string Name {
            get { return FunctionName; }
        }
        public Type ResultType(params Type[] operands) {
            return typeof(bool);
        }
        #endregion
    }
    // DateTime & Mixed
    public class IsWeekendFunction : ICustomFunctionDisplayAttributes {
        public const string FunctionName = "IsWeekend";
        static readonly IsWeekendFunction Instance = new IsWeekendFunction();
        IsWeekendFunction() { }
        //
        public static void Register() {
            CriteriaOperator.RegisterCustomFunction(Instance);
        }
        public static bool Unregister() {
            return CriteriaOperator.UnregisterCustomFunction(Instance);
        }
        #region ICustomFunctionOperatorBrowsable Members
        public FunctionCategory Category {
            get { return FunctionCategory.DateTime; }
        }
        public string Description {
            get { return "Determines if a day falls on a weekend."; }
        }
        public bool IsValidOperandCount(int count) {
            return count == 1;
        }
        public bool IsValidOperandType(int operandIndex, int operandCount, Type type) {
            return type == typeof(DateTime) || type == typeof(DateTime?);
        }
        public int MaxOperandCount {
            get { return 1; }
        }
        public int MinOperandCount {
            get { return 1; }
        }
        #endregion
        #region ICustomFunctionDisplayAttributes
        public string DisplayName {
            get { return "Is Weekend"; }
        }
        public object Image {
            get { return Properties.Resources.Actions_Calendar; }
        }
        #endregion
        #region ICustomFunctionOperator Members
        public object Evaluate(params object[] operands) {
            DateTime date = Convert.ToDateTime(operands[0]);
            return date.DayOfWeek == DayOfWeek.Sunday || date.DayOfWeek == DayOfWeek.Saturday;
        }
        public string Name {
            get { return FunctionName; }
        }
        public Type ResultType(params Type[] operands) {
            return typeof(bool);
        }
        #endregion
    }
    public class DaysBeforeOrAfterTodayFunction : ICustomFunctionDisplayAttributesEx {
        public const string FunctionName = "DaysBeforeOrAfterToday";
        static readonly DaysBeforeOrAfterTodayFunction Instance = new DaysBeforeOrAfterTodayFunction();
        DaysBeforeOrAfterTodayFunction() { }
        //
        public static void Register() {
            CriteriaOperator.RegisterCustomFunction(Instance);
        }
        public static bool Unregister() {
            return CriteriaOperator.UnregisterCustomFunction(Instance);
        }
        #region ICustomFunctionOperatorBrowsable Members
        public FunctionCategory Category {
            get { return FunctionCategory.DateTime; }
        }
        public string Description {
            get { return "Shows items that within X days before or after today."; }
        }
        public bool IsValidOperandCount(int count) {
            return count == 2;
        }
        public bool IsValidOperandType(int operandIndex, int operandCount, Type type) {
            return
                (operandIndex == 0 && type == typeof(DateTime)) ||
                (operandIndex == 1 && type == typeof(int));
        }
        public int MaxOperandCount {
            get { return 2; }
        }
        public int MinOperandCount {
            get { return 2; }
        }
        #endregion
        #region ICustomFunctionDisplayAttributes
        public string DisplayName {
            get { return "Days before or after today"; }
        }
        public object Image {
            get { return "SwitchTimeScalesTo;Size16x16;Colored"; }
        }
        #endregion
        #region ICustomFunctionOperator Members
        public object Evaluate(params object[] operands) {
            DateTime dt = Convert.ToDateTime(operands[0]);
            int days = Convert.ToInt32(operands[1]);
            DateTime start = DateTime.Today.AddDays(-days);
            DateTime end = DateTime.Today.AddDays(days);
            return (dt >= start) && (dt <= end);
        }
        public string Name {
            get { return FunctionName; }
        }
        public Type ResultType(params Type[] operands) {
            return typeof(bool);
        }
        public string GetOperandDisplayValue(object value, string displayText) {
            if(value is int)
                return string.Format("{0} days", value);
            return value?.ToString();
        }
        #endregion
    }
    // String
    public class NotBeginsWithFunction : ICustomFunctionDisplayAttributes {
        public const string FunctionName = "NotBeginsWith";
        static readonly NotBeginsWithFunction Instance = new NotBeginsWithFunction();
        NotBeginsWithFunction() { }
        //
        public static void Register() {
            CriteriaOperator.RegisterCustomFunction(Instance);
        }
        public static bool Unregister() {
            return CriteriaOperator.UnregisterCustomFunction(Instance);
        }
        #region ICustomFunctionOperatorBrowsable Members
        public FunctionCategory Category {
            get { return FunctionCategory.Text; }
        }
        public string Description {
            get { return "Selects items that do not start with the specified string."; }
        }
        public bool IsValidOperandCount(int count) {
            return count == 2;
        }
        public bool IsValidOperandType(int operandIndex, int operandCount, Type type) {
            return type == typeof(string);
        }
        public int MaxOperandCount {
            get { return 2; }
        }
        public int MinOperandCount {
            get { return 2; }
        }
        #endregion
        #region ICustomFunctionDisplayAttributes
        public string DisplayName {
            get { return "Not Begins With"; }
        }
        public object Image {
            get { return null; }
        }
        #endregion
        #region ICustomFunctionOperator Members
        public object Evaluate(params object[] operands) {
            if(operands[0] != null && operands[1] != null) {
                string str1 = operands[0].ToString(); string str2 = operands[1].ToString();
                return !str1.StartsWith(str2, StringComparison.InvariantCultureIgnoreCase);
            }
            return false;
        }
        public string Name {
            get { return FunctionName; }
        }
        public Type ResultType(params Type[] operands) {
            return typeof(bool);
        }
        #endregion
    }
}

File: WinForms/VB/GridMainDemo/Modules/Custom Functions.vb
Imports System
Imports DevExpress.Data.Filtering
Imports DevExpress.Data.Summary

Namespace DevExpress.XtraGrid.Demos

    ' Integer
    Public Class IsBlackFridayDiscountFunction
        Implements ICustomFunctionDisplayAttributes

        Public Const FunctionName As String = "IsBlackFridayDiscount"

        Private Shared ReadOnly Instance As IsBlackFridayDiscountFunction = New IsBlackFridayDiscountFunction()

        Private Sub New()
        End Sub

        '
        Public Shared Sub Register()
            CriteriaOperator.RegisterCustomFunction(Instance)
        End Sub

        Public Shared Function Unregister() As Boolean
            Return CriteriaOperator.UnregisterCustomFunction(Instance)
        End Function

#Region "ICustomFunctionOperatorBrowsable Members"
        Public ReadOnly Property Category As FunctionCategory Implements ICustomFunctionOperatorBrowsable.Category
            Get
                Return FunctionCategory.Math
            End Get
        End Property

        Public ReadOnly Property Description As String Implements ICustomFunctionOperatorBrowsable.Description
            Get
                Return "The discount amount is 15% or more."
            End Get
        End Property

        Public Function IsValidOperandCount(ByVal count As Integer) As Boolean Implements ICustomFunctionOperatorBrowsable.IsValidOperandCount
            Return count = 1
        End Function

        Public Function IsValidOperandType(ByVal operandIndex As Integer, ByVal operandCount As Integer, ByVal type As Type) As Boolean Implements ICustomFunctionOperatorBrowsable.IsValidOperandType
            Return SummaryItemTypeHelper.IsNumericalType(type)
        End Function

        Public ReadOnly Property MaxOperandCount As Integer Implements ICustomFunctionOperatorBrowsable.MaxOperandCount
            Get
                Return 1
            End Get
        End Property

        Public ReadOnly Property MinOperandCount As Integer Implements ICustomFunctionOperatorBrowsable.MinOperandCount
            Get
                Return 1
            End Get
        End Property

#End Region
#Region "ICustomFunctionDisplayAttributes"
        Public ReadOnly Property DisplayName As String Implements ICustomFunctionDisplayAttributes.DisplayName
            Get
                Return "Is Black Friday Discount"
            End Get
        End Property

        Public ReadOnly Property Image As Object Implements ICustomFunctionDisplayAttributes.Image
            Get
                Return Properties.Resources.isBlackFridayDiscount
            End Get
        End Property

#End Region
#Region "ICustomFunctionOperator Members"
        Public Function Evaluate(ParamArray operands As Object()) As Object Implements ICustomFunctionOperator.Evaluate
            Dim discount As Double = Convert.ToDouble(operands(0))
            Return discount >= 0.15
        End Function

        Public ReadOnly Property Name As String Implements ICustomFunctionOperator.Name
            Get
                Return FunctionName
            End Get
        End Property

        Public Function ResultType(ParamArray operands As Type()) As Type Implements ICustomFunctionOperator.ResultType
            Return GetType(Boolean)
        End Function
#End Region
    End Class

    ' DateTime & Mixed
    Public Class IsWeekendFunction
        Implements ICustomFunctionDisplayAttributes

        Public Const FunctionName As String = "IsWeekend"

        Private Shared ReadOnly Instance As IsWeekendFunction = New IsWeekendFunction()

        Private Sub New()
        End Sub

        '
        Public Shared Sub Register()
            CriteriaOperator.RegisterCustomFunction(Instance)
        End Sub

        Public Shared Function Unregister() As Boolean
            Return CriteriaOperator.UnregisterCustomFunction(Instance)
        End Function

#Region "ICustomFunctionOperatorBrowsable Members"
        Public ReadOnly Property Category As FunctionCategory Implements ICustomFunctionOperatorBrowsable.Category
            Get
                Return FunctionCategory.DateTime
            End Get
        End Property

        Public ReadOnly Property Description As String Implements ICustomFunctionOperatorBrowsable.Description
            Get
                Return "Determines if a day falls on a weekend."
            End Get
        End Property

        Public Function IsValidOperandCount(ByVal count As Integer) As Boolean Implements ICustomFunctionOperatorBrowsable.IsValidOperandCount
            Return count = 1
        End Function

        Public Function IsValidOperandType(ByVal operandIndex As Integer, ByVal operandCount As Integer, ByVal type As Type) As Boolean Implements ICustomFunctionOperatorBrowsable.IsValidOperandType
            Return type Is GetType(Date) OrElse type Is GetType(Date?)
        End Function

        Public ReadOnly Property MaxOperandCount As Integer Implements ICustomFunctionOperatorBrowsable.MaxOperandCount
            Get
                Return 1
            End Get
        End Property

        Public ReadOnly Property MinOperandCount As Integer Implements ICustomFunctionOperatorBrowsable.MinOperandCount
            Get
                Return 1
            End Get
        End Property

#End Region
#Region "ICustomFunctionDisplayAttributes"
        Public ReadOnly Property DisplayName As String Implements ICustomFunctionDisplayAttributes.DisplayName
            Get
                Return "Is Weekend"
            End Get
        End Property

        Public ReadOnly Property Image As Object Implements ICustomFunctionDisplayAttributes.Image
            Get
                Return Properties.Resources.Actions_Calendar
            End Get
        End Property

#End Region
#Region "ICustomFunctionOperator Members"
        Public Function Evaluate(ParamArray operands As Object()) As Object Implements ICustomFunctionOperator.Evaluate
            Dim [date] As Date = Convert.ToDateTime(operands(0))
            Return [date].DayOfWeek = DayOfWeek.Sunday OrElse [date].DayOfWeek = DayOfWeek.Saturday
        End Function

        Public ReadOnly Property Name As String Implements ICustomFunctionOperator.Name
            Get
                Return FunctionName
            End Get
        End Property

        Public Function ResultType(ParamArray operands As Type()) As Type Implements ICustomFunctionOperator.ResultType
            Return GetType(Boolean)
        End Function
#End Region
    End Class

    Public Class DaysBeforeOrAfterTodayFunction
        Implements ICustomFunctionDisplayAttributesEx

        Public Const FunctionName As String = "DaysBeforeOrAfterToday"

        Private Shared ReadOnly Instance As DaysBeforeOrAfterTodayFunction = New DaysBeforeOrAfterTodayFunction()

        Private Sub New()
        End Sub

        '
        Public Shared Sub Register()
            CriteriaOperator.RegisterCustomFunction(Instance)
        End Sub

        Public Shared Function Unregister() As Boolean
            Return CriteriaOperator.UnregisterCustomFunction(Instance)
        End Function

#Region "ICustomFunctionOperatorBrowsable Members"
        Public ReadOnly Property Category As FunctionCategory Implements ICustomFunctionOperatorBrowsable.Category
            Get
                Return FunctionCategory.DateTime
            End Get
        End Property

        Public ReadOnly Property Description As String Implements ICustomFunctionOperatorBrowsable.Description
            Get
                Return "Shows items that within X days before or after today."
            End Get
        End Property

        Public Function IsValidOperandCount(ByVal count As Integer) As Boolean Implements ICustomFunctionOperatorBrowsable.IsValidOperandCount
            Return count = 2
        End Function

        Public Function IsValidOperandType(ByVal operandIndex As Integer, ByVal operandCount As Integer, ByVal type As Type) As Boolean Implements ICustomFunctionOperatorBrowsable.IsValidOperandType
            Return operandIndex = 0 AndAlso type Is GetType(Date) OrElse operandIndex = 1 AndAlso type Is GetType(Integer)
        End Function

        Public ReadOnly Property MaxOperandCount As Integer Implements ICustomFunctionOperatorBrowsable.MaxOperandCount
            Get
                Return 2
            End Get
        End Property

        Public ReadOnly Property MinOperandCount As Integer Implements ICustomFunctionOperatorBrowsable.MinOperandCount
            Get
                Return 2
            End Get
        End Property

#End Region
#Region "ICustomFunctionDisplayAttributes"
        Public ReadOnly Property DisplayName As String Implements ICustomFunctionDisplayAttributes.DisplayName
            Get
                Return "Days before or after today"
            End Get
        End Property

        Public ReadOnly Property Image As Object Implements ICustomFunctionDisplayAttributes.Image
            Get
                Return "SwitchTimeScalesTo;Size16x16;Colored"
            End Get
        End Property

#End Region
#Region "ICustomFunctionOperator Members"
        Public Function Evaluate(ParamArray operands As Object()) As Object Implements ICustomFunctionOperator.Evaluate
            Dim dt As Date = Convert.ToDateTime(operands(0))
            Dim days As Integer = Convert.ToInt32(operands(1))
            Dim start As Date = Date.Today.AddDays(-days)
            Dim [end] As Date = Date.Today.AddDays(days)
            Return dt >= start AndAlso dt <= [end]
        End Function

        Public ReadOnly Property Name As String Implements ICustomFunctionOperator.Name
            Get
                Return FunctionName
            End Get
        End Property

        Public Function ResultType(ParamArray operands As Type()) As Type Implements ICustomFunctionOperator.ResultType
            Return GetType(Boolean)
        End Function

        Public Function GetOperandDisplayValue(ByVal value As Object, ByVal displayText As String) As String Implements ICustomFunctionDisplayAttributesEx.GetOperandDisplayValue
            If TypeOf value Is Integer Then Return String.Format("{0} days", value)
            Return value?.ToString()
        End Function
#End Region
    End Class

    ' String
    Public Class NotBeginsWithFunction
        Implements ICustomFunctionDisplayAttributes

        Public Const FunctionName As String = "NotBeginsWith"

        Private Shared ReadOnly Instance As NotBeginsWithFunction = New NotBeginsWithFunction()

        Private Sub New()
        End Sub

        '
        Public Shared Sub Register()
            CriteriaOperator.RegisterCustomFunction(Instance)
        End Sub

        Public Shared Function Unregister() As Boolean
            Return CriteriaOperator.UnregisterCustomFunction(Instance)
        End Function

#Region "ICustomFunctionOperatorBrowsable Members"
        Public ReadOnly Property Category As FunctionCategory Implements ICustomFunctionOperatorBrowsable.Category
            Get
                Return FunctionCategory.Text
            End Get
        End Property

        Public ReadOnly Property Description As String Implements ICustomFunctionOperatorBrowsable.Description
            Get
                Return "Selects items that do not start with the specified string."
            End Get
        End Property

        Public Function IsValidOperandCount(ByVal count As Integer) As Boolean Implements ICustomFunctionOperatorBrowsable.IsValidOperandCount
            Return count = 2
        End Function

        Public Function IsValidOperandType(ByVal operandIndex As Integer, ByVal operandCount As Integer, ByVal type As Type) As Boolean Implements ICustomFunctionOperatorBrowsable.IsValidOperandType
            Return type Is GetType(String)
        End Function

        Public ReadOnly Property MaxOperandCount As Integer Implements ICustomFunctionOperatorBrowsable.MaxOperandCount
            Get
                Return 2
            End Get
        End Property

        Public ReadOnly Property MinOperandCount As Integer Implements ICustomFunctionOperatorBrowsable.MinOperandCount
            Get
                Return 2
            End Get
        End Property

#End Region
#Region "ICustomFunctionDisplayAttributes"
        Public ReadOnly Property DisplayName As String Implements ICustomFunctionDisplayAttributes.DisplayName
            Get
                Return "Not Begins With"
            End Get
        End Property

        Public ReadOnly Property Image As Object Implements ICustomFunctionDisplayAttributes.Image
            Get
                Return Nothing
            End Get
        End Property

#End Region
#Region "ICustomFunctionOperator Members"
        Public Function Evaluate(ParamArray operands As Object()) As Object Implements ICustomFunctionOperator.Evaluate
            If operands(0) IsNot Nothing AndAlso operands(1) IsNot Nothing Then
                Dim str1 As String = operands(0).ToString()
                Dim str2 As String = operands(1).ToString()
                Return Not str1.StartsWith(str2, StringComparison.InvariantCultureIgnoreCase)
            End If

            Return False
        End Function

        Public ReadOnly Property Name As String Implements ICustomFunctionOperator.Name
            Get
                Return FunctionName
            End Get
        End Property

        Public Function ResultType(ParamArray operands As Type()) As Type Implements ICustomFunctionOperator.ResultType
            Return GetType(Boolean)
        End Function
#End Region
    End Class
End Namespace
